docs/design/copy-tracking.md

```diff
@@ -0,0 +1,286 @@
+
+```shell
+jj cp $SRC $DEST [OPTIONS]
+jj mv $SRC $DEST [OPTIONS]
+```
+
+These commands will rewrite the target commit to reflect the given move/copy
+instructions in its tree, as well as recording the rewrites on the Commit
+object itself for backends that support it (for backends that do not,
+these copy records will be silently discarded).
+
+Flags for the first two commands will include:
+
+```
+-r/--revision
+    perform the copy or move at the specified revision
+    defaults to the working copy commit if unspecified
+-f
+    force overwrite the destination path
+--after
+    record the copy retroactively, without modifying the targeted commit tree
+--resolve
+    overwrite all previous copy intents for this $DEST
+--allow-ignore-copy
+    don't error if the backend doesn't support copy tracking
+--from REV
+    specify a commit id for the copy source that isn't the parent commit
+```
+
+For backends which do not support copy tracking, it will be an error to use
+`--after`, since this has no effect on anything and the user should know that.
+The `supports_copy_tracking()` trait method is used to determine this.
+
+An additional command is provided to deliberately discard copy info for a
+destination path, possibly as a means of resolving a conflict.
+
+```shell
+jj forget-cp $DEST [-r REV]
+```
+
+## Behavioral Changes
+
+### Rebase Changes
+
+A well known and thorny problem in Mercurial occurs in the following scenario:
+
+1.  Create a new file A
+1.  Create new commits on top that make changes to file A
+1.  Whoops, I should rename file A to B. Do so, amend the first commit.
+1.  Because the first commit created file A, there is no rename to record; it's changing to a commit that instead creates file B.
+1.  All child commits get sad on evolve
+
+In jj, we have an opportunity to fix this because all rebasing occurs atomically
+and transactionally within memory. The exact implementation of this is yet to be
+determined, but conceptually the following should produce desirable results:
+
+1.  Rebase commit A from parents [B] to parents [C]
+1.  Get copy records from [D]->[B] and [D]->[C], where [D] are the common ancestors of [B] and [C]
+1.  DescendantRebaser maintains an in-memory map of commits to extra copy info, which it may inject into (2). When squashing a rename of a newly created file into the commit that creates that file, DescendentRebase will return this rename for all rebases of descendants of the newly modified commit. The rename lives ephemerally in memory and has no persistence after the rebase completes.
+1.  A to-be-determined algorithm diffs the copy records between [D]->[B] and [D]->[C] in order to make changes to the rebased commit. This results in edits to renamed files being propagated to those renamed files, and avoiding conflicts on the deletion of their sources. A copy/move may also be undone in this way; abandoning a commit which renames A->B should move all descendant edits of B back into A.
```

Comment:
nit: It sounds like this is mostly generic and not related to the problem mentioned above. I think it would be clearer to describe the general rebase behavior first and then describe how we can handle the thorny problem after that.

Reply:
Re-ordered this section.